src/transformers/modeling_blenderbot.py

```diff
@@ -0,0 +1,582 @@
+        :param static_key_val: Boolean, True if the key and value are held constant during decoding
+        :return: (attention tensor, new incremental_state)
+        """
+        assert mask is not None, "mask is None, please specify a mask"
+        batch_size, query_len, dim = query.size()
+        assert (dim == self.dim), "Dimensions does not match, query dimension is {} and configured dimension {}".format(dim, self.dim)
+        n_heads = self.n_heads
+        dim_per_head = dim // n_heads
+        #print(query.size())
+        if key is None and value is None:
+            key = value = query
+            _, key_len,dim=query.size()
+        
+        elif value is None:
+            value = key
+        assert key is not None
+        #print(key.size())
+        _, key_len, dim = key.size()
+        q = reshape_head(self.q_lin(query), n_heads, dim_per_head)
+        k = reshape_head(self.k_lin(key), n_heads, dim_per_head)
+        v = reshape_head(self.v_lin(value), n_heads, dim_per_head)
+        if incremental_state is None:
+            incremental_state = {}
+        if 'prev_value' in incremental_state:
+            prev_value = incremental_state['prev_value'].view(batch_size * n_heads, -1, dim_per_head)
+            if static_key_val:
+                v = prev_value
+            else:
+                v = torch.cat([prev_value, v], dim=1)
+        if 'prev_key' in incremental_state:
+            prev_key = incremental_state['prev_key'].view(batch_size * n_heads, -1, dim_per_head)
+            if static_key_val:
+                k = prev_key
+            else:
+                k = torch.cat([prev_key, k], dim=1)
+        if 'prev_mask' in incremental_state:
+            if static_key_val:
+                mask = incremental_state['prev_mask']
+            else:
+                mask = torch.cat([incremental_state['prev_mask'], mask], dim=2)  # key_len dimension
+        scale = math.sqrt(dim_per_head)
+        dot_prod = q.div_(scale).bmm(k.transpose(1, 2))
+        key_len = k.size(1)
+        attn_mask = (mask == 0).view(batch_size, 1, -1, key_len).repeat(1, n_heads, 1, 1) \
+            .expand(batch_size, n_heads, query_len, key_len).view(batch_size * n_heads, query_len, key_len)
+        assert attn_mask.shape == dot_prod.shape
+        dot_prod.masked_fill(attn_mask, neginf(dot_prod.dtype))
+        attn_weight = F.softmax(dot_prod, dim=-1, dtype=torch.float).type_as(query)
+        attn_weight = self.att_dropout(attn_weight)
+        attention = attn_weight.bmm(v)
+        attention = (attention.type_as(query)).view(batch_size, n_heads, query_len, dim_per_head).transpose(1, 2) \
+            .contiguous().view(batch_size, query_len, dim)
+        output = self.out_lin(attention)
+        new_incremental_state = {
+            'prev_key': k.view(batch_size, n_heads, -1, dim_per_head),
+            'prev_value': v.view(batch_size, n_heads, -1, dim_per_head),
+            'prev_mask': mask
```

Comment:
Do we have to save the `mask` as well? This is usually not done - instead the full mask is used every time even for cached past key, value states. It would be nice to not cache it here for consistency with other models IMO (GPT2, T5)